lib/systems/parse.nix: mkSkeletonFromList: improve readability

amjoseph-nixpkgs opened the pull request:
### Motivation

The main purpose of this PR is to make the basis for `mkSkeletonFromList`'s decision between `cpu-kernel-libcabi` vs `cpu-vendor-os` clear, without changing its behavior.  The existing code obscures this decision behind a sequence of prioritized matches (i.e. `if-then`) which jump around between different coordinates.

Two side benefits of this PR:

1. It makes the root cause of #165836 obvious: we are missing a case for `cpu-vendor-libcabi`.  This is why nixpkgs stumbles over `*-none-*`.

2. It illuminates some very weird corner cases in the existing logic, like `*-${vendor}-ghcjs` overriding the `vendor` field, and `mingw32` being transformed into `windows` in some cases.

### Instructions for verifying no behavior change

```
curl -L https://raw.githubusercontent.com/amjoseph-nixpkgs/nixpkgs/a9595374806477c40eab88acbc06ca7d641364ea/lib/systems/test.nix > lib/systems/test.nix
nix eval --raw -f lib/systems/test.nix | egrep -ve '-apple-(mingw32|eabi|elf|ghcjs)' > after
git switch --detach origin/master
nix eval --raw -f lib/systems/test.nix | egrep -ve '-apple-(mingw32|eabi|elf|ghcjs)' > before
diff -u before after
```

The diff should produce no output.  You should verify that the `*-apple-*` tuples removed by the `egrep -v` filters are nonsensical (changing the behavior in these nonsensical cases allowed simplification -- see below for details).  Here is how [`lib/systems/test.nix`](https://github.com/amjoseph-nixpkgs/nixpkgs/blob/a9595374806477c40eab88acbc06ca7d641364ea/lib/systems/test.nix) was created:

1. Copy-and-paste of the old (unsimplified) version of `mkSkeletonFromList`
2. Comment it out
3. Form a list consisting of one or more attrsets for each possible coordinate value which could participate in triggering each of the `if` clauses in the original code
4. Append "default values" for each of the three coordinates.
5. Run `mkSkeletonFromList` on the `cartesianProductOfSets` of the attrsets.

Note that if `vendor` is missing from the return value of `mkSkeletonFromList`, [`lib/systems/test.nix` will set `vendor="unknown"`](https://github.com/amjoseph-nixpkgs/nixpkgs/blob/1682f2b5a896e4a26edd9844dddc184d897ab4d4/lib/systems/test.nix#L84)



###### Things done

- Built on platform(s)
  - [x] x86_64-linux
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Commit:
lib/systems/parse.nix: mkSkeletonFromList: improve readability

The main purpose of this PR is to make the basis for
`mkSkeletonFromList`'s decision between `cpu-kernel-libcabi` vs
`cpu-vendor-os` clear, without changing its behavior.  The existing
code obscures this decision behind a sequence of prioritized matches
(i.e. `if-then`) which jump around between different coordinates.

Two side benefits of this PR:

1. It makes the root cause of #165836 obvious: we are missing a case
   for `cpu-vendor-libcabi`.  This is why nixpkgs stumbles over
   `*-none-*`.

2. It illuminates some very weird corner cases in the existing
   logic, like `*-${vendor}-ghcjs` overriding the `vendor` field,
   and `mingw32` being transformed into `windows` in some cases.

Co-authored-by: John Ericson <git@JohnEricson.me>

amjoseph-nixpkgs commented:
I need to run some test vectors through this to double-check that I haven't changed the behavior of `mkSkeletonFromList`.  Does anybody know a good source for a big pile of autoconf-names?

amjoseph-nixpkgs commented:
Good to go.  For ease of reviewing, see the copypasta at the top of the first post.

amjoseph-nixpkgs commented:
@sternenseemann should be added as a reviewer here.

amjoseph-nixpkgs commented:
@sternenseemann and @Ericson2314, have you had a chance to look at this yet?

There are a few PRs open (like #186484) which, if merged ahead of this one, will invalidate the proof that the behavior of `mkSystemFromString` is unchanged.

amjoseph-nixpkgs commented:
Ping @sternenseemann @Ericson2314 @alyssais @nbp

I want to submit other changes which will create a merge conflict with this PR.  What should I do?

We're starting to run into a lot more problems with calling Musl an ABI.  Nixpkgs needs understand that "big-endian 64-bit powerpc on Musl always uses the ELFv2 ABI (see #182807)", but since nixpkgs thinks that Musl is an ABI there is no way to express this other than special-purpose hackery like #191436.  The fact that `{build,host,target}Platform` don't have separate fields for libc and abi which can be matched independently is the main obstacle.  Fixing that will create a merge conflict with this PR.

I've radically shortened the description at the top of this PR in the hope of attracting some kind of review.

Ericson2314 commented:
@amjoseph-nixpkgs this looks good! Sorry I didn't see it for so long, Please harass me on Matrix in #cross-compiling:nixos.org to make sure this stuff gets reviewed, this is not the first time you made a good thing that went neglected by us!

Ericson2314 commented on lib/systems/parse.nix:
```diff
@@ -407,27 +407,29 @@ rec {
+        cpu    = elemAt l 0;
+        kernel = elemAt l 1;
+        abi    = elemAt l 2;
+      } // lib.optionalAttrs (elemAt l 1 != "linux") {
```
```suggestion
```
I think you discovered a bug :) I don't see why that shouldn't just be unconditional. Still, you did such a meticulus job trying to preserve behavior I am fine if you do this behavior in a subsequent PR.

amjoseph-nixpkgs replied:
I looked this over pretty carefully and am convinced it has no detectable effect, so I've included it in the PR.

The output of `mkSkeletonFromList` is always passed to `mkSystemFromSkeleton`, which will insert a `vendor.unknown` if the vendor field is missing and `!isDarwin && !isWindows`, which will always be the case if `(elemAt l 1 == "linux")`.

So removing the conditional has no visible effect, and is easier to read.

sternenseemann reviewed (approved):
Also did some half-elaborate sanity checking of my own which seems to check out.

One thing is that some things in `parse.nix` have changed since July, but I'm not sure if we need to adjust any code in here necessarily.

Ericson2314 commented on lib/systems/parse.nix:
```diff
@@ -407,27 +407,28 @@ rec {
+              hasPrefix "genode" (elemAt l 2)
+      then {
+        cpu    = elemAt l 0;
+        vendor = if elemAt l 2 == "ghcjs" then "unknown" else elemAt l 1;
```
```suggestion
        vendor = elemAt l 1;
```
The config in `lib/systems/examples.nix` is `js-unknown-ghcjs` so I think this is fine and the original is a silly copy paste thing from the line above.

amjoseph-nixpkgs commented:
Merge conflict due to #82131.

amjoseph-nixpkgs commented:
Squashed, because it wasn't worth rebasing past 25 separate commits.

I've added `*-*-freebsd*` to the `cpu-vendor-os` fork of the `*-*-*` case.

Note that freebsd can be used in the other (`cpu-kernel-environment`) fork as well; `x86_64-freebsd-gnu` [is still valid](https://www.debian.org/ports/kfreebsd-gnu/) although I don't think glibc still builds on freebsd.  But there's no need to parse it unless somebody is serious about adding that capability.

Ericson2314 commented:
Sorry we didn't merge before! :(

amjoseph-nixpkgs commented:
> Git push to origin failed for release-22.11 with exitcode 1

I think Sterni fixed this in #202405.  If not, or if I need to do anything more, ping me.
